pkg/blueprint/firstboot_customizations.go

```diff
@@ -0,0 +1,173 @@
+	CACerts []string `json:"cacerts,omitempty" toml:"cacerts,omitempty"`
+
+	// Registration command as generated by the Satellite server. Required, if
+	// type is set to "satellite".
+	Command string `json:"command,omitempty" toml:"command,omitempty"`
+}
+
+// AAPFirstbootCustomization contains fields specific to AAP firstboot
+// customizations.
+type AAPFirstbootCustomization struct {
+	FirstbootCommonCustomization
+
+	// Optional CA certificate to enroll into the system before executing the
+	// firstboot script.
+	CACerts []string `json:"cacerts,omitempty" toml:"cacerts,omitempty"`
+
+	// Job template URL as generated by the AAP server. Required if type is set
+	// to "aap". Example URLs are
+	// https://aap.example.com/api/controller/v2/job_templates/9/callback/ or
+	// https://aap.example.com/api/v2/job_templates/9/callback/ depending on the
+	// AAP version.
+	JobTemplateURL string `json:"job_template_url,omitempty" toml:"job_template_url,omitempty"`
+
+	// The host config key. Required if type is set to "aap".
+	HostConfigKey string `json:"host_config_key,omitempty" toml:"host_config_key,omitempty"`
+}
+
+func (t FirstbootScriptCustomization) MarshalJSON() ([]byte, error) {
+	b, err := t.union.MarshalJSON()
+	return b, err
+}
+
+func (t *FirstbootScriptCustomization) UnmarshalJSON(b []byte) error {
+	err := t.union.UnmarshalJSON(b)
+	return err
+}
+
+func (t FirstbootScriptCustomization) MarshalTOML() ([]byte, error) {
+	b, err := t.union.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	return jsonToToml(b)
+}
+
+func (t *FirstbootScriptCustomization) UnmarshalTOML(data any) error {
+	return unmarshalTOMLviaJSON(t, data)
+}
+
+var ErrMissingCustomContents = errors.New("missing contents field for custom firstboot customization")
+var ErrUnknownFirstbootCustomization = errors.New("unknown firstboot customization: missing or invalid type field")
+var ErrMissingAAPFields = errors.New("missing job_template_url or host_config_key field for aap firstboot customization")
+var ErrMissingSatelliteCommand = errors.New("missing command field for satellite firstboot customization")
+
+// SelectUnion returns the specific firstboot customization types. The detection
+// is based on the "type" field in the JSON payload. Returns nil for irrelevant types,
+// or if the type is unknown or if there is an error during unmarshaling.
+func (sp FirstbootScriptCustomization) SelectUnion() (*CustomFirstbootCustomization, *SatelliteFirstbootCustomization, *AAPFirstbootCustomization, error) {
+	var fcc FirstbootCommonCustomization
+	err := json.Unmarshal(sp.union, &fcc)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	switch strings.ToLower(fcc.Type) {
+	case "custom":
+		var cc CustomFirstbootCustomization
+		jd := json.NewDecoder(bytes.NewReader(sp.union))
+		jd.DisallowUnknownFields()
+		err = jd.Decode(&cc)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		if cc.Contents == "" {
+			return nil, nil, nil, ErrMissingCustomContents
+		}
```

Comment:
So I was thinking if we would not want some kind of helper function for this, but then, we cannot use cc, sc and ac as the same parameter. If you came up with an idea, feel free to do it, I just wanted to put the thought here.

```
func decodeAndValidate(data json.RawMessage) error {
    jd := json.NewDecoder(bytes.NewReader(data))
    jd.DisallowUnknownFields()
    return jd.Decode(**xxxx**)
}

```